File: 121_BestTimeToBuyAndSellStock/BestTimeToBuyANdSellStock.cc
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        if (prices.size() == 0) {
            return 0;
        }
        int minVal = prices[0];
        int best_profit = 0;
        for (auto price : prices) {
            best_profit = std::max(best_profit, price - minVal);
            minVal = std::min(minVal, price);
        }
        return best_profit;
    }
};

File: 121-best-time-to-buy-and-sell-stock/best-time-to-buy-and-sell-stock.js
// Say you have an array for which the ith element is the price of a given stock on day i.
//
// If you were only permitted to complete at most one transaction (i.e., buy one and sell one share of the stock), design an algorithm to find the maximum profit.
//
// Note that you cannot sell a stock before you buy one.
//
// Example 1:
//
//
// Input: [7,1,5,3,6,4]
// Output: 5
// Explanation: Buy on day 2 (price = 1) and sell on day 5 (price = 6), profit = 6-1 = 5.
//              Not 7-1 = 6, as selling price needs to be larger than buying price.
//
//
// Example 2:
//
//
// Input: [7,6,4,3,1]
// Output: 0
// Explanation: In this case, no transaction is done, i.e. max profit = 0.
//
//


/**
 * @param {number[]} prices
 * @return {number}
 */
var maxProfit = function(prices) {
    var min = prices[0];
    var curr = 0;
    for (var i = 1;i < prices.length;++i) {
	curr = Math.max(curr, prices[i] - min);
	min = Math.min(min, prices[i]);
    }
    return curr;
};

